### ISSUE TITLE
[1708.02002] Focal Loss for Dense Object Detection [paper-reading]

### ISSUE THREAD

**Issue by yoheikikuta** [yoheikikuta/paper-reading#19]
## 論文リンク
https://arxiv.org/abs/1708.02002

## 公開日（yyyy/mm/dd）
2017/08/07

## 概要
Focal Loss を導入して RetinaNet という one stage の高性能の object detection model を作ったという話。
one stage detector は objectness を測る直接の仕組みがないため大量の anchor に対して loss を計算するが、classification を考えるとこの anchor のうちほとんどが背景のような簡単な sample になっている。しかしそのような簡単な sample でも数が多いので total loss に占める割合は大きくなってしまい、興味ある sample の loss が埋もれてしまうという問題がある。
そこで、モデルの出力確率値が大きいものに関しては cross entropy の値が小さくなるような factor をかける Focal Loss を提案し、効率的に学習がなされるようにして、two stage detector をも上回るような結果を出した。Focal Loss は $` - (1-p_t)^γ \ln(p_t) `$ のような形。
モデルの中身は backbone network から特徴量を抽出して Feature Pyramid Network を構築し、それぞれの pyramid の特徴量で classification と bbox regression を実施するもの。

**Comment by yoheikikuta** [comment]
最近の object detection 系の話をちゃんとフォローしてなかったので、その辺の知識をつけるため。
どこかで概要を見かけて、 loss function の設定に工夫があるみたいなのにちょっと興味があるので、ちゃんと自分で読んでみることにした。

**Comment by yoheikikuta** [comment]
まず、two stage detector と one stage detector を整理しておく。

- two stage detector
  object を detect するステージと detect した object を classify するステージに分かれているもの
  [Faster R-CNN](https://arxiv.org/abs/1506.01497) に代表されるモデルで、object か否かを anchor 毎に分類する Region Proposal Network(RPN) が一つ目のステージを担い、その後 RoI pooling を経て分類するという二つ目のステージが続く（色々はしょりまくってるが）。
  これの二つ目のステージを segmentaion のマスクを予測するものを追加して拡張したのが [Mask R-CNN](https://arxiv.org/abs/1703.06870) である。
  two stage detector は RPN と classification を分けてがっつり学習してるので性能は高いが、その分遅いという問題がある。
- one stage detector
  上述の object の detect と classify をまとめてやってしまおうというもの。
  代表的なモデルは  [YOLO](https://arxiv.org/abs/1804.02767) とか [SSD](https://arxiv.org/abs/1512.02325) など。
  具体的な話には立ち入らないが、two stage のように RPN を賢いものになるように学習して、というプロセスがないので、基本的にはガッとばらまいた anchor のうち、ちゃんとした物体ぽいっものを選ぶために classification のスコアが高いものをフィルタしたりする。RPN のように objectness そのものを表す量がないのでこういう方法に頼る必要がある。
  その分速度は早くなるが性能は落ちるという問題がある。

**Comment by yoheikikuta** [comment]
今回の RetinaNet は one stage detector だ。

上記の性質を考えた時に問題となる点として、「大量の anchor に対して classification をして loss を算出するが、特に何も写ってないような easy な anchor が多くて、そいつらによって loss が dominate されるので意味ある loss の計算にならない」という点が挙げられる。

ちょっと ↑ の書き方だけだと分かりづらいかもしれないので、もう少し具体的な想像をしておく。

例えば anchor が 100 個あるときの loss を考える。
anchor それぞれの cross entropy の和が loss になるわけだが、このうち 90 個は全然 object を写してないただの背景の場合にどうなるだろうか。
残りの 10 個がちゃんと object が写っているものでこれは分類が難しいのでモデルの出力確率値が低くて（例えば 0.2）、 10 * (- ln(0.2)) = 16 くらい loss に寄与するとする。
一方で 90 個の方は簡単なのでモデルの確率出力値が高い（例えば 0.8）としよう。この場合 loss への寄与は 90 * (- ln(0.8)) = 20 くらいになる。こっちの方が高くなってしまう。
普通に anchor を作るとさらに比は悪くなって 1:1000 とかにもなるので、そこまで意味のない anchor に対する loss のせいで学習が我々の意図したものと食い違ってしまう。

先行研究では難易度が高い anchor を集めてくる Online Hard Example Mining などの手法が提案されていた。

この論文では、それと逆の発想をして、簡単な anchor の loss への寄与を抑えるように loss function を変形するだけで目的を達成しようというものだ。

**Comment by yoheikikuta** [comment]
ということでこの論文の主題。
classification の loss を通常の Cross Entropy (CE) から Focal Loss (FL) に変更するというものである。
なんと $` - (1-p_t)^γ \ln(p_t) `$ として、p が 1 に近い場合に値が小さくなるように γ の値を大きくする、というシンプルなものである。
ここで、p_t は二値分類の y=1 のときが p で y=-1 のときが (1-p) という定義である。

<p align="center">
  <img src="https://i.imgur.com/0pbhQC3.png">
</p>

γ の値が大きい場合に CE との相対的な差が大きくなって、 easy anchor の loss への寄与がより小さくなっていることが見て取れる。

これによって、簡単な classification の寄与が抑えられるので、興味ある anchor の classification をより効率的に学習していけるという算段である。

**Comment by yoheikikuta** [comment]
より tuning した形として overall factor として overall factor を入れた次の形を採用している。
α_t は y=1 のときが α で y=-1 のときが 1-α を取る。

<p align="center">
  <img src="https://i.imgur.com/I9ycwc1.png">
</p>

この α は Focall Loss の意味ではそこまで重要でなく、これを導入すると少し性能がよくなるという程度のもの。

パラメタの値としては γ=2, α=0.25 を採用している。

classification のときは、各 anchor に対して class 毎に sigmoid で確率を計算して Focal Loss を計算して和をとる、というのを

**Comment by yoheikikuta** [comment]
モデルの詳細は以下の要素が分かっていればよく、しかもそれらは従来から知られているものである。
- 特徴抽出のための backbone モデル（ResNet など）
- bbox の算出をマルチスケールの情報を使って実施するための Feature Pyramid Network (FPN)
- classification と box regression の構造

<p align="center">
  <img src="https://i.imgur.com/osrEEk5.png">
</p>

この図だけだとモデルの詳細としては色々落ちている部分がある。

あとで書く。

**Comment by yoheikikuta** [comment]
しかしこうして書いてみるとちゃんと理解しようと思ったら結構大変だよな。

自分は昔 Faster R-CNN をしっかり読んだのですんなり理解できたが、object detection を勉強しようという人がいきなりこの手の論文から着手すると理解するまでに結構頑張る必要がありそう。

まあ object detection の必須構成要素なので、一回ちゃんと理解すれば two stage detector の Mask R-CNN とかもそこまで苦労せずに理解できると思う（自分はまだちゃんと読んでないけど、眺めた感じそう思った）。

**Comment by yoheikikuta** [comment]
備忘のためのメモを残しておく。

上述のように backbone network から C3, C4, C5 を取得しているが、実装を見るとその部分がちゃんと取得できているのか理解するまでにちょっと苦戦したのでそのメモ。

- [C3, C4, C5 を取得しているところ](https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/models/retinanet.py#L276)
- [backbone が resnet のときに返している output](https://github.com/broadinstitute/keras-resnet/blob/master/keras_resnet/models/_2d.py#L74-L91) これは別レポジトリ
  これは block の塊が 4 つあってそれを全部返しているので C2, C3, C4, C5 を返している
- [retinanet で C3, C4, C5 だけを取得するように設定しているところ](https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/models/resnet.py#L112)
  この outputs[1:] を見落としててちょっと混乱した

**Comment by yoheikikuta** [comment]
あとは実験結果。
いくつか実験をしているが、抜粋。

- Focal Loss のパラメタ α と γ を変えた時の振る舞い（学習が MS COCO trainval35k で test が minval）

<p align="center">
  <img src="https://i.imgur.com/VXqxX72.png">
</p>

- 各種モデルとの比較

<p align="center">
  <img src="https://i.imgur.com/s4hzZ4i.png">
</p>

- inference time と性能の関係

<p align="center">
  <img src="https://i.imgur.com/iWojQej.png">
</p>

**Comment by yoheikikuta** [comment]
性能はなかなか良くて、one stage detector の中では最上位と思っていいだろう。
自分で学習させて使ってみた感触としては、YOLOv3 と同等以上かなという感触である。
